festim/problem.py

```diff
@@ -0,0 +1,164 @@
+import dolfinx
+from dolfinx import fem
+from dolfinx.nls.petsc import NewtonSolver
+import basix
+import ufl
+from mpi4py import MPI
+import numpy as np
+import tqdm.autonotebook
+import festim as F
+
+
+class Problem:
+    def __init__(
+        self,
+        mesh=None,
+        sources=None,
+        exports=None,
+        subdomains=None,
+        boundary_conditions=None,
+        settings=None,
+    ) -> None:
+        self.mesh = mesh
+        # for arguments to initialise as empty list
+        # if arg not None, assign arg, else assign empty list
+        self.subdomains = subdomains or []
+        self.boundary_conditions = boundary_conditions or []
+        self.sources = sources or []
+        self.exports = exports or []
+        self.settings = settings
+
+        self.dx = None
+        self.ds = None
+        self.function_space = None
+        self.facet_meshtags = None
+        self.volume_meshtags = None
+        self.formulation = None
+        self.bc_forms = []
+
+    @property
+    def volume_subdomains(self):
+        return [s for s in self.subdomains if isinstance(s, F.VolumeSubdomain)]
+
+    @property
+    def surface_subdomains(self):
+        return [s for s in self.subdomains if isinstance(s, F.SurfaceSubdomain)]
+
+    def define_function_spaces(self):
+        """Creates the function space of the model, creates a mixed element if
+        model is multispecies. Creates the main solution and previous solution
+        function u and u_n. Create global DG function spaces of degree 0 and 1
+        for the global diffusion coefficient"""
+
+        degree = 1
+        element_CG = basix.ufl.element(
+            basix.ElementFamily.P,
+            self.mesh.mesh.basix_cell(),
+            degree,
+            basix.LagrangeVariant.equispaced,
+        )
+
+        return element_CG
```

Comment:
I understand this is shared functionality across classes. However this method doesn't do what it says in the docstrings (ie. define a function space), it only returns a finite element. Several options:
- have a bit of code duplication in each child class
- convert this into a property `element_CG` so that classes can call `self.element_CG`

Note: it would be beneficial to expose `degree` to users one day. Could we make a note of this and/or a new issue?

Reply:
I think it's best to have a bit of code duplication then. As for exposing degree, I can do that in a small pull request after this, I'll add a note.